### DISCUSSION TITLE
Any luck using two Coral USB?

### DISCUSSION THREAD

**Discussion by t0ny-peng** [blakeblackshear/frigate#1351] [General]
I bought one Coral USB stick but the seller shipped two. And they said I can just keep it, free of charge. What a nice guy.

So the question is, does having two Coral USB bring any extra performance boost? Or is it even possible?

**Comment by blakeblackshear** [comment]
Needs to be configured, but it will perform even better assuming you have enough USB power/bandwidth. It will probably draw too much power for an RPi if thats what you are using.
https://blakeblackshear.github.io/frigate/configuration/detectors

**Reply by t0ny-peng** [reply]
Nice. Didn't see that part. I'm using NUC so it should the power consumption should be fine. I'll see how to mount them into PVE.

**Reply by code-in-progress** [reply]
I'm running both USB and mPCIe corals. Detection is SUPER fast (imho):

coral1 is my mPCIe and coral2 is the USB.

![image](https://user-images.githubusercontent.com/33837776/125125052-10ce4780-e0c7-11eb-9991-ea8d1011de95.png)

**Reply by t0ny-peng** [reply]
@code-in-progress Hmm. The speed you get is around 3 times faster than my Coral USB. I'm using NUC as host for PVE. Frigate is running as Docker container inside an Ubuntu virtual machine and Coral USB is mounted to the VM. Guess that's where the overhead comes from. But I'm pretty happy with inference speed around 25ms. 


Update: After running frigate in PVE host, the inference time reduced to 8.75ms. That made me even happier.

**Reply by asdkjasdfla** [reply]
How did you run it directly in PVE? Do you mean you installed it directly in ubuntu instead of using Docker in ubuntu?

**Reply by t0ny-peng** [reply]
@asdkjasdfla PVE is based on Debian so you can install Docker there easily.

**Comment by balajeek** [comment]
I just received one coral usb, I already setup frigate and using it for last 3 days and i can see the cpu runs max%, can't wait to install the coral USB.  How do i install on on my HP NUC intel core i5 with usb 3 ports running ubuntu latest version.

do i have to follow all the steps mentioned in the product doucment https://coral.ai/docs/accelerator/get-started/ 
and then set the detector in the config to use coral?

thanks.

**Reply by t0ny-peng** [reply]
@balajeek Are you running frigate within a Python environment or in a docker container?

In the second case, you need to mount the device into the container. IIRC it's `/dev/bus/usb`

**Reply by balajeek** [reply]
Yes its the docker and its in the docker-compose yaml like this, but still don't understand why cpu usage is that high.

    devices:
      - /dev/bus/usb:/dev/bus/usb

**Reply by balajeek** [reply]
This morning i tried to check the cpu% few times and it shows in the range of 65%,  Is this normal ?
I was under impression it should be less than 5% !

**Reply by balajeek** [reply]
Also since i have the HASS integration too, there is a sensor called **Coral Inference Speed** which is in the range of 10ms, that tells me coral is indeed processing, isn't !

**Reply by t0ny-peng** [reply]
@balajeek Only the TFLite part will be moved to Coral. The rest(motion detection, recording, etc.) still remain on your NUC. Maybe you have too many cameras, or camera with high resolution?

In my case, I have three 720P camera, and the CPU usage of Frigate(from `docker stats`) on my i5-5250U is around 25% when idle, and 60% when there's motion in the camera. 100% means one CPU core is fully loaded.

I would't worry too much since CPU is meant to be used..

**Reply by balajeek** [reply]
I see, thanks for the feedback.
i use 3 1080p cameras.

**Comment by allenrong88** [comment]
Any further configuration is required apart from the code mentioned in the documentation? I have got my code listed below, which is essentially copied from the doc. Got two Coral USB connected to my system recently and both got recognized without any issues. The inference speed was around 10ms for one Coral USB, I was hoping to have a better speed considering double of the capabilities. But still both show around 10ms without any improvement. No other changes have been made to the cameras. Any tips will be appreciated. Thanks.

detectors:
  coral1:
    type: edgetpu
    device: usb:0
  coral2:
    type: edgetpu
    device: usb:1

**Reply by Sch0field9** [reply]
One possibility is that you are limited by the bandwith of the usb. Check your motherboard manual and make sure that the corals are not on a usb port from the same usb controller. It is also possible that you are limited by your memory speed.

**Reply by allenrong88** [reply]
Thanks for that. I am not a hardware guy and not exactly sure whether they are sharing the same USB controller. I am using HP Microserver G8, purposely one was plugged into the USB3.0 port and the other was on the USB2.0, which I assume they are sitting on the different controller. The manual does not mention it, unfortunately :(
